### ISSUE TITLE
Use `Liquid::Drop` as a method to interface with `Liquid::Template#render`

### ISSUE THREAD

**Issue by ronaldtse** [lutaml/lutaml#74]
From Claude-3.5:

To pass a Ruby object that responds to methods instead of a hash into `template.render(...)`, you can use Liquid's `Drop` class. Here's how you can do it:

1. First, create a custom class that inherits from `Liquid::Drop`:

```ruby
class MyDrop < Liquid::Drop
  def initialize(obj)
    @obj = obj
  end

  def liquid_method_missing(method)
    @obj.send(method) if @obj.respond_to?(method)
  end
end
```

2. Then, create your Ruby object with the desired methods:

```ruby
class MyObject
  def x
    1
  end

  def z
    OpenStruct.new(a: 2)
  end
end
```

3. Now you can use it like this:

```ruby
template = Liquid::Template.parse("{{x}} {{z.a}}")
my_object = MyObject.new
result = template.render(MyDrop.new(my_object), strict_variables: true)
puts result  # Output: 1 2
```

This approach allows you to use a Ruby object with methods instead of a hash. The `MyDrop` class acts as a wrapper around your object, intercepting method calls from the Liquid template and forwarding them to your object if it responds to those methods.

The `liquid_method_missing` method in `MyDrop` is called for any method that isn't explicitly defined in the Drop class. It checks if the wrapped object responds to the method and calls it if it does.

This solution maintains the safety of Liquid templates while allowing you to use Ruby objects with methods instead of just hashes.

**Comment by ronaldtse** [comment]
Seems that all return values of the Drop class needs to be either:
* Drop
* Hash
* Array
* String
* Numeric

Perhaps the `liquid_method_missing` method is what we use do to do that (we can mix that into Shale::Mapper), to extend any class to return one of these objects.

e.g. we can have a `to_liquid`:
```ruby
class Xmi > Shale::Mapper
  def liquid_method_missing(method)
    return nil unless self.respond_to?(method)

    value = self.send(method) 
    return value if value.class < [Liquid::Drop, Array, Hash, String, Numeric]

    wrap_with_drop(value)
  end
end
```

----

It's possible to create a more dynamic approach that automatically generates a Drop class based on the structure of your Ruby object. Here's a method to do this:

```ruby
require 'liquid'

def create_drop_class(obj)
  Class.new(Liquid::Drop) do
    def initialize(obj)
      @obj = obj
    end

    obj.methods(false).each do |method|
      define_method(method) do
        result = @obj.send(method)
        if result.is_a?(Hash) || result.is_a?(Array) || result.is_a?(String) || result.is_a?(Numeric)
          result
        else
          create_drop_class(result).new(result)
        end
      end
    end

    def liquid_method_missing(method)
      @obj.send(method) if @obj.respond_to?(method)
    end
  end
end
```

This `create_drop_class` method does the following:

1. It creates a new class that inherits from `Liquid::Drop`.
2. It defines an initializer that takes the original object.
3. It iterates over all the methods defined directly on the object (excluding inherited methods).
4. For each method, it defines a corresponding method in the Drop class.
5. If the result of calling a method is a basic type (Hash, Array, String, or Numeric), it returns it directly.
6. If the result is another object, it recursively creates a Drop class for that object.
7. It also defines a `liquid_method_missing` method to handle any methods not explicitly defined.

You can use it like this:

```ruby
class MyObject
  def x
    1
  end

  def y
    "hello"
  end

  def z
    OpenStruct.new(a: 2, b: "world")
  end
end

my_object = MyObject.new
MyObjectDrop = create_drop_class(my_object)

template = Liquid::Template.parse("{{x}} {{y}} {{z.a}} {{z.b}}")
result = template.render(MyObjectDrop.new(my_object), strict_variables: true)
puts result  # Output: 1 hello 2 world
```

This approach has several advantages:

1. It's more dynamic and can handle complex nested structures.
2. It automatically creates Drop classes for all methods in your object.
3. It maintains the safety of Liquid by only exposing methods that exist on your object.
4. It handles nested objects by creating Drop classes for them as well.

Keep in mind that this approach creates a new class for each unique object structure. If you're working with many different object types, you might want to cache these generated Drop classes to avoid creating too many classes dynamically.

Would you like me to explain any part of this solution in more detail or provide any additional information?

**Comment by ronaldtse** [comment]
Seems that all return values of the Drop class needs to be either:
* Drop
* Hash
* Array
* String
* Numeric

Perhaps the `liquid_method_missing` method is what we use do to do that (we can mix that into Shale::Mapper), to extend any class to return one of these objects.

e.g. we can have a `to_liquid`:
```ruby
class Xmi > Shale::Mapper
  def liquid_method_missing(method)
    return nil unless self.respond_to?(method)

    value = self.send(method) 
    return value if value.class < [Liquid::Drop, Array, Hash, String, Numeric]

    wrap_with_drop(value)
  end
end
```

**Comment by kwkwan** [comment]
Closed as completed via b5f0e86bff40c4d63e0e6d922aed4e8f58742bf6
